### PR TITLE
Disable '..' in query string for artifact URI

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/10653?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10653/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10653
```

</p>
</details>

### What changes are proposed in this pull request?
Solving path traversal vulnerability.

#### Problem
When users create experiment with base artifact URI like `http://foo/bar`, MLflow appends `/{{run_id}}/artifact` sub path to it when saving/reading each run's artifact, such as `http://foo/bar/{{run_id}}/artifact". Every GET request for artifacts are validated so that the requested path are under this path, effectively prevents attackers to read any files outside the that run directory.

However, there is one hack to bypass this, which is query string. When MLflow appends `/{{run_id}}/artifact`, it will be inserted **before** query string of the specified artifact root. For example, if the artifact root is `http://foo/bar?a=a`, the run's artifact URI will be ``http://foo/bar{{run_id}}/artifact?a=a`.

This allows path traversal by adding malformed query string like "../../../../etc", which results in run's artifact location to be ``http://foo/bar{{run_id}}/artifact?../../../../etc``, which is then resolved to `/etc` as a local path.

#### Solution
This PR resolves this by explicitly validating query string passed as a part of artifact URI. It simply check if the query string contains ".." or not (with decoding).

There were some alternatives considered:
- Ignore query string entirely when appending ``{{run_id}}/artifacts`` => I'm afraid that some cases we need query string to access artifact location e.g. S3 ARN can have a region as query string.
- Remove all string after "?" when resolving local path. => "?" is a valid character in some OS.

Currently users can specify any query strings to artifact URI, for example, "http:///?/../path".

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Validated the fix prevents experiment creation with malformed query string:
```
curl -X POST -H 'Content-Type: application/json' -d '{"name": "poc", "artifact_location": "http:///??/../../../../../../../../../../../../../../etc/"}' 'http://127.0.0.1:5000/ajax-api/2.0/mlflow/experiments/create'
```
```
{"error_code": "INVALID_PARAMETER_VALUE", "message": "Invalid query string"}
```

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/uri.py [hunk 1]
```diff
@@ -447,3 +452,20 @@ def validate_path_is_safe(path):
...
+        raise MlflowException("Invalid query string", error_code=INVALID_PARAMETER_VALUE)
+
+
+def _decode(url):
+    # Keep decoding until the url stops changing
+    while True:
+        decoded = urllib.parse.unquote(url)
+        if decoded == url:
+            return url
+        url = decoded
```
Comment (harupy) [inline mlflow/utils/uri.py:471]: /ai any chance that this function goes into an infinite loop?

Comment (harupy) [reply mlflow/utils/uri.py:471]: /ai Can we use recursion instead?

**Review by serena-ruan — APPROVED** [review]
LGTM!

File: mlflow/utils/uri.py (hunk 1 above)
Comment (serena-ruan) [reply mlflow/utils/uri.py:471]: I feel like 10 is enough 🤔  Who would encode a path to such complex stuff?

Comment (B-Step62) [reply mlflow/utils/uri.py:471]: Makes sense, will limit it to 10:)

Comment (harupy) [reply mlflow/utils/uri.py]: If we reach 10, I think that's some malicious URI.

**Review by harupy — APPROVED** [review]
LGTM

**Pull request merged**